src/CMakeLists.txt

```diff
@@ -42,10 +42,49 @@ target_link_libraries(DLAF
 
 target_compile_features(DLAF PUBLIC cxx_std_14)
 
-target_compile_definitions(DLAF PUBLIC DLAF_SOURCE_LOCATION=__func__)
-
 target_add_warnings(DLAF)
 
+# ----- Options, Flags, Defines, ...
+# DLAF_FUNCTION_NAME
+include(CheckCXXSourceCompiles)
+check_cxx_source_compiles("int main() { const char *name = __PRETTY_FUNCTION__; }" is_pretty_function_available)
+
+target_compile_definitions(DLAF PUBLIC
+  DLAF_FUNCTION_NAME=$<IF:$<BOOL:is_pretty_function_available>,__PRETTY_FUNCTION__,__func__>
+)
+
+# Assertions
+if (CMAKE_BUILD_TYPE STREQUAL Debug)
+  set(DLAF_ASSERT_DEFAULT_LOW  ON)
+  set(DLAF_ASSERT_DEFAULT_MED  ON)
+  set(DLAF_ASSERT_DEFAULT_HIGH ON)
+else()
+  set(DLAF_ASSERT_DEFAULT_LOW  ON)
+  set(DLAF_ASSERT_DEFAULT_MED  OFF)
+  set(DLAF_ASSERT_DEFAULT_HIGH OFF)
+endif()
+
+# DLAF_ASSERT_DEFAULT_BUILDTYPE makes persistent the build type for which the current set of flags is configured
+# Checking its value, it is possible to reset to flag values when build type is changed
+# Build type comparison is case sensitive (Debug != Release != RelWithDebInfo != release)
+#
+# In order to force a flag configuration, you have to:
+# - set DLAF_ASSERT_DEFAULT_BUILDTYPE=${CMAKE_BUILD_TYPE} (in order to bypass this check)
+# - set DLAF_ASSERT_ENABLE_* variable(s) as you prefer
+if (NOT DLAF_ASSERT_DEFAULT_BUILDTYPE STREQUAL CMAKE_BUILD_TYPE)
+  set(DLAF_ASSERT_DEFAULT_BUILDTYPE ${CMAKE_BUILD_TYPE} CACHE INTERNAL "" FORCE)
+
+  set(DLAF_ASSERT_ENABLE_LOW  ${DLAF_ASSERT_DEFAULT_LOW}  CACHE BOOL "Enable low impact assertions"    FORCE)
+  set(DLAF_ASSERT_ENABLE_MED  ${DLAF_ASSERT_DEFAULT_MED}  CACHE BOOL "Enable medium impact assertions" FORCE)
+  set(DLAF_ASSERT_ENABLE_HIGH ${DLAF_ASSERT_DEFAULT_HIGH} CACHE BOOL "Enable high impact assertions"   FORCE)
+endif()
```

Comment:
@teonnik this is the solution I was talking about.

- It gives you a default value for debug or release builds
- If you do not change the build type, you are allowed to customize your flags setup
- As soon as you switch build type, flags configuration is reset to default one
- You can define a custom configuration for flags at first project configuration (e.g. for CI particular builds) as described in the documentation

Reply:
Yes, it is the where I started from.

IMHO it is a trade-off between having the extra functionality and the cost of the additional variable. One may interpret the `DLAF_ASSERT_DEFAULT_BUILDTYPE` like a variable saying that you are going to specify flags setup for a particular build. However, it is true that in certain situations this auto-reset functionality may be counter-productive, but I think it may not the "main stream" situation.

About your comment, I'm absolutely not against it, we just have to decide to loose the "requirement" of not resetting flags on build type changes. In that case the default setup works just on the first configuration (with respect to the first build type chosen), then it is up to the developer to turn ON/OFF flags if he decides to switch build type. But it is completely ok for me, even if it will happen that I will forget to set assert switches 😂

Reply:
After some discussion with Alberto, the following question came up:

What happens if `DLAF_ASSERT_DEFAULT_BUILDTYPE=Release` and I change from `CMAKE_BUILD_TYPE=Release` to `Debug` but forget to update the former?

The answer is that changing `DLAF_ASSERT_ENABLE` variables from the CACHE will no longer have an affect as the variables will always be overwritten to the default.

What happens if I change both `DLAF_ASSERT_DEFAULT_BUILDTYPE` and `CMAKE_BUILD_TYPE` from `Debug` to `Release`?

Unfortunately, the CACHE variables `DLAF_ASSERT_ENABLE` remain with values from `Default` as the `IF` condition was not triggered.

To avoid these traps, we decided to stick with the simple solution and just provide a single set of defaults via the `option()` clauses.